### DISCUSSION TITLE
Continuous Plasma Beam Emission in PICMI

### DISCUSSION THREAD

**Discussion by Aquios7** [BLAST-WarpX/warpx#4556] [Q&A] [answered]
I'm trying to produce a continuous plasma beam.

I was trying to use UniformFluxDistribution in PICMI, but I can't get it to run. I haven't been able to find any examples, and the only information I've seen for it is in the source code for pywarpx.

Has this been implemented yet for the PICMI warpx code? Since it wasn't on the PICMI parameters section of the readme, I don't know if I can even implement it.

This was my attempt at getting an electron beam to emit particles from the left side of the simulation, but it crashes immediately on startup. I'm assuming I'm missing parameters, but I don't know what else I would need to add.

# electron beam
electron_beam = picmi.UniformFluxDistribution(lower_bound=[-dX, None, -dZ],
                                              upper_bound=[0, None, dZ],
                                              directed_velocity=[V, 0., 0.],
                                              rms_velocity=[rms_vel_e]*3,
                                              flux=1,
                                              flux_normal_axis="x",
                                              surface_flux_position=[-dX, None, -dZ],
                                              flux_direction=1
                                              )

If there is a better way to implement an constant stream of particles using PICMI, then I'd happily add that to me simulation. Otherwise I may have to remake my python file into an input file and run using the nfluxpercell for emission.

Thank You for any help.

**Comment by RemiLehe** [comment, marked as answer]
Hi, 
We are in the process of adding an example that uses `UniformFluxDistribution`, as part of this PR: 
https://github.com/ECP-WarpX/WarpX/pull/4413/files#diff-a3f76b976bcab620a76e40fb49a17b2f8458a078331fe99893aa504b164b140fR197
Does this help?

Also: do get any error message, when your simulation crashes at startup?

**Comment by Aquios7** [comment]
Thank you for the quick reply. I'm currently trying to get a small example working from the example you gave me, but I wasn't able to get it working immediately on Windows. Something with mpi4py is not working with my conda distribution of warpX.

About the error message, here is what I run:
(warpx) D:\warpx_folder>mpiexec /np 1 python UniformFluxDistributionTest.py

And here is what I get:

OMP initialized with 32 OMP threads
AMReX (23.07) initialized
Traceback (most recent call last):
  File "D:\warpx_folder\UniformFluxDistributionTest.py", line 192, in <module>
    sim.initialize_warpx()
  File "C:\~\anaconda3\envs\warpx\Lib\site-packages\pywarpx\picmi.py", line 1843, in initialize_warpx
    pywarpx.warpx.init(mpi_comm, max_step=self.max_steps, stop_time=self.max_time)
  File "C:\~\anaconda3\envs\warpx\Lib\site-packages\pywarpx\WarpX.py", line 95, in init
    libwarpx.initialize(argv, mpi_comm=mpi_comm)
  File "C:\~\anaconda3\envs\warpx\Lib\site-packages\pywarpx\_libwarpx.py", line 456, in initialize
    self.libwarpx_so.warpx_init()
OSError: [WinError -529697949] Windows Error 0xe06d7363

I had assumed that I was using the UniformFluxDistribution wrong, and that the error was a result of incorrect parameters.

**Reply by Aquios7** [reply]
An update about this. I had written this reply last week, but didn't send it until today, but I have gotten a beam running in my simulation. The example helped me figure out what parts of the UniformFluxDensity I was missing, and now I'm working on adjusting the density of the beam itself. I'm going to mark this as solved, but I can answer any questions on the issues I was having if needed.
Thank You.

**Reply by ax3l** [reply]
That sounds great. If you have more questions on this setup, let us know here.
